_posts/fastapi-rust-5-rocket-0.5.md

```diff
@@ -0,0 +1,212 @@
+
+At this point I was thoroughly lost, if I needed to pass data to something, and some trait outside of my control was requiring it to be `'static`, then I couldn’t pass references right? I didn’t want to clone all the data for every request, as doing so would be very inefficient. Even if I wanted to do that, some of the data I needed to pass was not `Clone`. I decided to go out on a limb and ask for help. I’ll admit I was afraid, as much as I try to contribute positively to the open source community, the responses I’ve received have been very mixed. However, luckily, one of the contributors to the `lamedh_` crates came through with the answer!
+
+I had heard of, though never used, `std::sync::Arc` before. I’m not really qualified to explain how this works, but I’ll do my best to convey _my_ understanding. `std::rc::Rc` effectively lets you have shared, read-only ownership of an object. If you do something like this:
+
+```rust
+use std::rc::Rc;
+
+let original = String::from("blah");
+let first = Rc::new(original);
+let second = Rc::clone(&first);
+```
+
+Then you effectively end up with two different owned values `first` and `second` which can be dereferenced to read data as if it were `original` (note that creating `first` took ownership of `original` so you can’t access that one anymore). Because the values are owned, you’re not passing references, and therefore don’t have to worry about lifetimes!
+
+This isn’t magic, `Rc` stands for “reference counting” which is a common way of keeping track of memory in more dynamic languages (e.g. Python). Behind the scenes, Rust is keeping track of how many times `Rc::clone` has been used for a given piece of data and once they’re _all_ dropped, it will free the memory. So even if `Handler` goes out of scope and gets dropped with a `Future` still on the loose, the data it was given is still perfectly valid!
+
+“Wait a second, you said you were using `Arc` though?!” Yep! `Arc` is just the version of `Rc` that is `Send`, meaning it is safe in threaded contexts. Because we’re using Tokio and Tokio can use threads to execute `Future`s, everything has to be `Send` and therefore we use `Arc`! And thus the problem of sharing data in Tokio futures is solved!
+
+### Step 2.1: Being Less Selfish
+
+In order to aid me with `Arc`ing the codebase, I pulled out most of the methods on `RocketHandler` into functions instead which took just the data they needed rather than the whole `self`. This made it _way_ easier to reason about the code, and understanding led me to faster, more confident modifications. I prefer this style anyway because it makes it clear when you _call_ something exactly what that something is taking. An object-oriented “method” style hides whether something is just reading data, or mutating it, or outright consuming it until you jump to definition. Why not take full advantage of how explicit Rust is? Here's an example to illustrate my point:
```

Comment:
```suggestion
In order to aid me with `Arc`ing the codebase, I pulled out most of the methods on `RocketHandler` into functions which took just the data they needed rather than the whole `self`. This made it _way_ easier to reason about the code and make faster, more confident modifications. I prefer this style anyway because it makes it clear when you _call_ something exactly what that something is taking. An object-oriented “method” style hides whether something is just reading data, or mutating it, or outright consuming it until you jump to definition. Why not take full advantage of how explicit Rust is? Here's an example to illustrate my point:
```